### PR TITLE
GH-6170: Search bar: the chosen word from the drop-down is not shown

### LINKED ISSUE #6170: Search bar: the chosen word from the drop-down is not shown
**Steps to Reproduce:**
1. Go to UI -> click on search bar to perform a search
2. Start to type any word (e.g: manage) to get the full word hint in the drop-down
3. Choose any option from the drop-down hints (e.g: management)
image

4. Pay attention on the search results and on the chosen word in search bar

**Observed Results:**
The chosen word from the drop-down hint (e.g: management) is not shown on the search bar, however the search result is performed
image

**Expected Results:**
The chosen word from the drop-down hint (e.g: management) is shown on the search bar

### PR DESCRIPTION
**Pull request by benmarcotte**
Closes #6170

### CODE REVIEW

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]
Don't think we need the pipe nor searching through the focused element when clicking to find if it exist in the array.

What you can do is look into ViewChild and get the selected suggestion value from there, in which you can use that value to populate the input with

----
The pipe is throwing the error right away when the application is just open in console
The pipe throws error when clicking a value as well

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 1]
```diff
@@ -46,6 +46,7 @@ export class SearchBoxComponent {
    * for example when we click inside the search result section.
    */
   private ignoreCloseEvent = false;
+  public chosenWord = '';
```
Comment (saad-mo) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: This should be private, per our coding guidelines - https://sap.github.io/spartacus-docs/coding-guidelines/#services

**Review by saad-mo — CHANGES_REQUESTED** [review]
This PR addresses the issue at hand, but also introduces a small UI bug as well. When you navigate away from the search results page, the key is still retained in the search box. The search box should be cleared when you navigate away from the search results page. 

Update: After reviewing the behavior on dev branch, this is a pre-existing UI bug and it's best to address in this PR.

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 2]
```diff
@@ -56,8 +58,17 @@ export class SearchBoxComponent {
     protected searchBoxComponentService: SearchBoxComponentService,
     @Optional()
     protected componentData: CmsComponentData<CmsSearchBoxComponent>,
-    protected winRef: WindowRef
-  ) {}
+    protected winRef: WindowRef,
+    protected router: Router
+  ) {
+    this.router.events.subscribe((event: Event) => {
```
Comment (saad-mo) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: We need to `unsubscribe` from this observable in `ngOnDestroy()`

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts (hunk 1 above)
Comment (saad-mo) [reply projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: since this member needs to be accessible in the template, it cannot be private.

**Review by saad-mo — CHANGES_REQUESTED** [review]
- We will need unit test to make sure the chosen word in the search box is same as in the URL
- Also another unit test to check if the search box is empty when not in the search results view

**Review by saad-mo — DISMISSED** [review]
Looks good, thanks!

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 3]
```diff
@@ -56,8 +62,20 @@ export class SearchBoxComponent {
     protected searchBoxComponentService: SearchBoxComponentService,
     @Optional()
     protected componentData: CmsComponentData<CmsSearchBoxComponent>,
-    protected winRef: WindowRef
-  ) {}
+    protected winRef: WindowRef,
+    protected routingService: RoutingService,
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: breaking change

will have a call with you soon

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 4]
```diff
@@ -56,8 +62,20 @@ export class SearchBoxComponent {
...
+    protected routingConfigService: RoutingConfigService
+  ) {
+    this.updateChosenWordFromUrl = this.routingService
+      .getRouterState()
+      .subscribe((data) => {
+        const search = this.routingConfigService
+          .getRouteConfig('search')
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: we can do something better than the logic applied here (will discuss together in a call in relation to the breaking change)

Basically, look at the param query and context of the page through the routingService instead to reset the chosen word

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts (hunk 1 above)
Comment (bgambocjaviniar) [reply projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: let's omit the public keyword, however do note that if it needed to be private, you can always use getters/setters

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 5]
```diff
@@ -46,7 +51,8 @@ export class SearchBoxComponent {
    * for example when we click inside the search result section.
    */
   private ignoreCloseEvent = false;
-
+  public chosenWord = '';
+  public updateChosenWordFromUrl: Subscription;
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: let's keep the variable naming consistent as it is throughout the application

change to `subscription`

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 6]
```diff
@@ -232,4 +254,8 @@ export class SearchBoxComponent {
       this.ignoreCloseEvent = false;
     });
   }
+
+  ngOnDestroy() {
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: I don't know how this is working, but it needs to implement OnDestroy

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts (hunk 5 above)
Comment (bgambocjaviniar) [reply projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: just putting the link to the above comment https://github.com/SAP/spartacus/pull/8298#discussion_r568719587 as you might miss it

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts [hunk 7]
```diff
@@ -7,12 +7,15 @@ import {
...
+  RoutingConfigService,
+  RoutingService,
 } from '@spartacus/core';
-import { Observable, of } from 'rxjs';
+import { BehaviorSubject, Observable, of } from 'rxjs';
 import { CmsComponentData } from '../../../cms-structure/page/model/cms-component-data';
 import { SearchBoxComponentService } from './search-box-component.service';
 import { SearchBoxComponent } from './search-box.component';
 import { SearchResults } from './search-box.model';
+import { RouterState } from './../../../../../core/src/routing/store/routing-state';
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts]: this should be included in the @spartacus/core import

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 8]
```diff
@@ -46,17 +52,35 @@ export class SearchBoxComponent {
...
 
+  constructor(
+    searchBoxComponentService: SearchBoxComponentService,
+    componentData: CmsComponentData<CmsSearchBoxComponent>,
+    winRef: WindowRef,
+    // tslint:disable-next-line: unified-signatures
+    routingService: RoutingService
+  );
+  /**
+   * @deprecated
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: As mentioned in the DM, it would be best to add comments, such as the constructor deprecation in `projects/storefrontlib/src/cms-components/navigation/search-box/search-box-component.service.ts`

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 9]
```diff
@@ -89,6 +113,15 @@ export class SearchBoxComponent {
     switchMap((config) => this.searchBoxComponentService.getResults(config))
   );
 
+  ngOnInit() {
+    this.subscription = this.routingService
+      .getRouterState()
+      .subscribe((data) => {
+        if (!data.nextState && data.state.params?.query !== this.chosenWord)
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: 1) Any reason to keep the `!data.nextState`? What's the use case for it?

2) As mentioned before, wouldn't it be better to include `data.state.context.type === PageType.CONTENT_PAGE && data.state.context.id === 'search'` in order to be sure that you're really in the search page. 
As I'm not sure if xxx.params?.query is the only use case for the search page. This property may exist in other places. (not sure, but it's a possibility)

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 10]
```diff
@@ -89,6 +113,15 @@ export class SearchBoxComponent {
     switchMap((config) => this.searchBoxComponentService.getResults(config))
   );
 
+  ngOnInit() {
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: let's explicitly add `:void`

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 11]
```diff
@@ -159,6 +192,10 @@ export class SearchBoxComponent {
     return <HTMLElement>this.winRef.document.activeElement;
   }
 
+  updateChosenWord(chosenWord: string) {
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: let's explicitly add `:void`

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts [hunk 12]
```diff
@@ -232,4 +269,8 @@ export class SearchBoxComponent {
       this.ignoreCloseEvent = false;
     });
   }
+
+  ngOnDestroy() {
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.ts]: let's explicitly add `:void`

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts [hunk 13]
```diff
@@ -262,6 +289,54 @@ describe('SearchBoxComponent', () => {
...
+          url: null,
+          queryParams: null,
+          params: { query: 'camera' },
+          context: null,
+          cmsRequired: null,
+        },
+        navigationId: null,
+      };
+      const input = fixture.debugElement.query(By.css('input'));
+      const PRODUCT_SEARCH_STRING = 'camera';
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts]: let's move this as 'global' as it's being used in 2 test cases

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts [hunk 14]
```diff
@@ -262,6 +289,54 @@ describe('SearchBoxComponent', () => {
       ).toBeTruthy();
     });
 
+    it('should contain chosen word from the dropdown', () => {
+      const mockRouterState = {
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts]: let's move this object as 'global' and setting the params property as an empty object to be used here https://github.com/SAP/spartacus/pull/8298/files#diff-8713f6e9795265679e836cf3c74e49434601908590027faf3be387cb6e4ee5d0R72

After, you can just re-define an object using that mockedObject and spread it with the value you have in this current params

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts [hunk 15]
```diff
@@ -262,6 +289,54 @@ describe('SearchBoxComponent', () => {
...
+      };
+      const input = fixture.debugElement.query(By.css('input'));
+      const PRODUCT_SEARCH_STRING = 'camera';
+      input.nativeElement.value = PRODUCT_SEARCH_STRING;
+      input.triggerEventHandler('keydown.enter', {});
+      routerState$.next(mockRouterState);
+      fixture.detectChanges();
+      expect(searchBoxComponent.chosenWord).toEqual(PRODUCT_SEARCH_STRING);
+      expect(
+        fixture.debugElement.query(By.css('input')).nativeElement.value
```
Comment (bgambocjaviniar) [inline projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts]: `fixture.....(by.css('input'))` can be replaced by `input`

**Review by bgambocjaviniar — CHANGES_REQUESTED** [review]

File: projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts (hunk 15 above)
Comment (bgambocjaviniar) [reply projects/storefrontlib/src/cms-components/navigation/search-box/search-box.component.spec.ts]: wasn't changed, but let's get on a call to address some of the new changes done

**Review by bgambocjaviniar — APPROVED** [review]
